### DISCUSSION TITLE
[adsk] Per Prim / UFE Xform Edit Locking

### DISCUSSION THREAD

**Discussion by nxkb** [Autodesk/maya-usd#2999] [Ideas]
Hello there,
We have a need to lock down the transforms above certain assets in our scenes so that the animation department does not unknowingly overwrite the Xform contributions from the layout department and wind up with a disagreement about where a character/prop should exist in the shot!

Using the AL shape, we would simply `cmds.lockNode(l=True)` the Al_usdmaya_Xform nodes to prevent the issue, but that is no longer possible now that we're dealing with ufe nodes.

What would be the best approach to get this functionality with the new adsk shape? Did I miss something obvious?

Some ideas:
1. Add Support for UFE nodes in lockNode command? This would be the most obvious, but I'm not sure that fits with the UFE design strategy not to mention that this feature could be a ways out if it requires a ufe change.
2. Add a special USD attribute that MayaUsd understands. There's a good post about per-prim edit permissions that would describe how this might work in an application: "[usd-interest] Per PrimSpec Edit Permission?"
3. Use a custom edit routing callback? Are these able to block edits or do they they just choose where they live? My only reservation with something custom is that "locking" seems fairly fundamental so I'd imagine each studio rolling their own slightly different version.

Let me know what you think...

**Comment by gelshoffNV** [comment]
Take a look at https://github.com/Autodesk/maya-usd/blob/681ae52e8a4126a97ea16d9b304a5c7835ad94f2/test/lib/ufe/testAttribute.py#L1640

You can set metadata on the ufe attr so it becomes locked.

maya-usd seems to read metadata from USD using this token
https://github.com/Autodesk/maya-usd/blob/cc4a36a32bb8baa840c8a4811bfd98b006acc536/lib/mayaUsd/base/tokens.h#L90

**Reply by nxkb** [reply]
Thanks for pointing me to those, very helpful. :)

> You can set metadata on the ufe attr so it becomes locked.

This is great, however it requires UFE 3.0 / Maya 2023 which we aren't on quite yet.

> maya-usd seems to read metadata from USD using this token

The "mayaLock" attribute metadata works pretty well, but It's worth noting that there are a couple sharp edges there. 
- Because this is an attribute metadata, you'll need to lock all 3 attributes separately
- If those attributes haven't been authored yet, this will error. So you need to create the attributes but make sure not to set a value and freeze any updates from coming in, before locking them down. 

So it could still be worthwhile to have some broader locking that wouldn't be so tied to individual attributes.
```
def Xform "transA" (
    mayaLock = "On"
)
{
```
Otherwise you're left with doing something like this:
```python
xformAPI = UsdGeom.XformCommonAPI(prim)
xformAPI.SetTranslate([0,0,0])
xformAPI.SetScale([1,1,1])
xformAPI.SetRotate([0,0,0])
prim.GetProperty('xformOp:translate').SetMetadata(mayaUsd.lib.MetadataTokens.Lock, mayaUsd.lib.Tokens.On)
prim.GetProperty('xformOp:translate').Clear()
prim.GetProperty('xformOp:scale').SetMetadata(mayaUsd.lib.MetadataTokens.Lock, mayaUsd.lib.Tokens.On)
prim.GetProperty('xformOp:scale').Clear()
prim.GetProperty('xformOp:rotateXYZ').SetMetadata(mayaUsd.lib.MetadataTokens.Lock, mayaUsd.lib.Tokens.On)
prim.GetProperty('xformOp:rotateXYZ').Clear()
```
instead of:
```python
prim.SetMetadata(mayaUsd.lib.MetadataTokens.Lock, mayaUsd.lib.Tokens.On)
```
